packages/accordion/src/useAccordion.ts

```diff
@@ -0,0 +1,172 @@
+      if (allowMultiple) {
+        // Just toggle the current one in the array
+        const newExpanded = Array.isArray(expandedItems) ? expandedItems : []
+
+        if (!checkIsExpanded({ key })) {
+          handleSetExpandedItemsChange([...newExpanded, key])
+        } else {
+          handleSetExpandedItemsChange(
+            newExpanded.filter(expandedItem => expandedItem !== key),
+          )
+        }
+      } else {
+        if (!checkIsExpanded({ key })) {
+          handleSetExpandedItemsChange([key])
+        } else {
+          handleSetExpandedItemsChange([])
+        }
+      }
+    },
+    [
+      checkIsExpanded,
+      handleSetExpandedItemsChange,
+      allowMultiple,
+      expandedItems,
+    ],
+  )
+
+  function registerAccordionHeader({
+    ref,
+    options,
+  }: {
+    ref: AccordionRef
+    options: RegisterOptions<KeyType>
+  }) {
+    if (options.role === 'accordionHeader') {
+      const accordionHead = ref
+      elementRefDictionary.current.accordionHeaders[options.key] = ref
+
+      const headerId = accordionHead.id
+      if (headerId === undefined || headerId === '') {
+        setAttributes(accordionHead, { id: `${options.key}` })
+      }
+    }
+
+    return {
+      ref,
+      options,
+    }
+  }
+
+  // Register the body
+  function registerAccordionBody({
+    ref,
+    options,
+  }: {
+    ref: AccordionRef
+    options: RegisterOptions<KeyType>
+  }) {
+    if (options.role === 'accordionBody') {
+      const accordionBody = ref
+      elementRefDictionary.current.accordionBodies[options.key] = ref
+
+      let accordionBodyId = accordionBody.id
+      if (accordionBodyId === undefined || accordionBodyId === '') {
+        accordionBodyId = options.key + '-Body'
+        setAttributes(accordionBody, { id: accordionBodyId })
+      }
+
+      // try to link up the aria controls/labelled by
+      const accordionHead =
+        elementRefDictionary.current.accordionHeaders[options.key]
+      if (accordionHead) {
+        const headerId = accordionHead.id
+        setAttributes(accordionHead, { 'aria-controls': accordionBodyId })
+        setAttributes(accordionBody, {
+          'aria-labelledby': headerId,
+          role: 'region',
+        })
+      }
+    }
+
+    return {
+      ref,
+      options,
+    }
+  }
+
+  function register(
+    ref: AccordionRef | null,
+    options: RegisterOptions<KeyType>,
+  ): {
+    ref: AccordionRef
+    options: RegisterOptions<KeyType>
+  } | null {
+    return (
+      // ref && registerAccordionHeader({ ref, options })
+      ref && registerAccordionBody(registerAccordionHeader({ ref, options }))
+    )
+  }
+
+  function getHeaderProps({ key }: { key: KeyType }) {
+    return {
+      onClick: () => toggleExpanded({ key }),
+      onKeyUp: (event: React.KeyboardEvent) => {
+        if (['Enter', ' '].includes(event.key)) {
+          toggleExpanded({ key })
+        }
+      },
+      'aria-expanded': checkIsExpanded({ key }),
+    }
+  }
```

Comment:
Should we wrap this with `useCallback`?

Reply:
I was trying to figure this out because it's really the event handlers that need to get memoized.  We could `useCallback` getHeaderProps but it'll return a new set of click/key handlers.

So I guess we could add some memoization so that the functions are always the same if the same key is provided: `useCallback(memoize(getHeaderProps(key))`.  Let's try it out.

Reply:
Unfortunately, it doesn't seem to help as much with the re-renders as splitting out the contents of the accordion body into it's own component.

Reply:
We can leave it for now.